cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -0,0 +1,232 @@
+import co.cask.cdap.internal.app.runtime.schedule.trigger.TriggerJsonDeserializer;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ScheduleId;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset that stores and oindexes proram schedules, so that they can be looked by their trigger keys.
+ */
+public class ProgramScheduleStoreDataset extends AbstractDataset {
+
+  private static final String SCHEDULE_COLUMN = "sch";
+  private static final byte[] SCHEDULE_COLUMN_BYTES = Bytes.toBytes(SCHEDULE_COLUMN);
+  private static final String TRIGGER_KEY_COLUMN = "tk"; // trigger key
+  private static final byte[] TRIGGER_KEY_COLUMN_BYTES = Bytes.toBytes(TRIGGER_KEY_COLUMN);
+
+  // package visible for the dataset definition
+  static final String EMBEDDED_TABLE_NAME = "it"; // indexed table
+  static final String INDEX_COLUMNS = TRIGGER_KEY_COLUMN; // trigger key
+
+  private static final Gson GSON = new GsonBuilder().registerTypeAdapter(Trigger.class, new TriggerJsonDeserializer()).create();
+
+  private final IndexedTable store;
+
+  public ProgramScheduleStoreDataset(DatasetSpecification spec,
+                                     @EmbeddedDataset(EMBEDDED_TABLE_NAME) IndexedTable store) {
+    super(spec.getName(), store);
+    this.store = store;
+  }
+
+  /**
+   * Add one or more schedules to the store.
+   *
+   * @param schedules the schedules to add
+   *
+   * @throws AlreadyExistsException if one of the schedules already exists
+   */
+  public void addSchedules(ProgramSchedule... schedules) throws AlreadyExistsException {
+    for (ProgramSchedule schedule : schedules) {
+      String scheduleKey = rowKeyForSchedule(schedule.getProgramId().getParent(), schedule.getName());
+      if (!store.get(new Get(scheduleKey)).isEmpty()) {
+        throw new AlreadyExistsException(schedule.getProgramId().getParent().schedule(schedule.getName()));
+      }
+      store.put(new Put(scheduleKey, SCHEDULE_COLUMN, GSON.toJson(schedule)));
+      int count = 0;
+      for (String triggerKey : extractTriggerKeys(schedule)) {
+        String triggerRowKey = rowKeyForTrigger(scheduleKey, count++);
+        store.put(new Put(triggerRowKey, TRIGGER_KEY_COLUMN, triggerKey));
+      }
+    }
+  }
+
+  /**
+   * Removes one or more schedules from the store. Succeeds whether the schedules exist or not.
+   *
+   * @param scheduleIds the schedules to delete
+   */
+  public void deleteSchedules(ScheduleId ... scheduleIds) {
+    for (ScheduleId scheduleId : scheduleIds) {
+      String scheduleKey = rowKeyForSchedule(scheduleId);
+      store.delete(new Delete(scheduleKey));
+      byte[] prefix = keyPrefixForTriggerScan(scheduleKey);
+      try (Scanner scanner = store.scan(new Scan(prefix, Bytes.stopKeyForPrefix(prefix)))) {
```

Comment:
I might be good to add a simple comment saying that all schedule's trigger keys are prefixed by the scheduleKey, hence scan and delete would remove the schedule and all triggers.